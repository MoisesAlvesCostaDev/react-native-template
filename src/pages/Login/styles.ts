import styled from 'styled-components/native';
import {BACKGROUND_COLOR_PRIMARY, BACKGROUND_COLOR_SECONDARY} from "assets/colors";



export const BodyContainer = styled.View`
  flex: 6;
  background-color: ${BACKGROUND_COLOR_SECONDARY};
  border-top-left-radius: 30px;
  border-top-right-radius: 30px;
  padding-top:20px;

`;
