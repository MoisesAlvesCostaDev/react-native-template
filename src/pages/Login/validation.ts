import * as Yup from "yup";

export const validationSchema = Yup.object().shape({
  email: Yup.string().email("E-mail inválido").required("O E-mail é obrigatório"),
  password: Yup.string().min(4,"Senha deve conter no mínimo 4 caracteres").required("Senha é obrigatória"),
});
