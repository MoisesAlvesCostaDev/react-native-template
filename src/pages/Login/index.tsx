import React, { useRef, useState } from "react";
import { Image, TouchableOpacity } from "react-native";
import {
  Screen,
  TitleScreen,
  HeaderContainer,
  ImageHeaderContainer,
} from "assets/styles/Screen";
import { BodyContainer, ForgotPasswordText } from "./styles";
import { Form } from "@unform/mobile";
import { FormHandles } from "@unform/core";
import TextInput from "components/forms/TextInput";
import PasswordInput from "components/forms/PasswordInput";
import Button from "components/Button";
import { Api } from "../../services/api";
import Modal from "components/Modal";
import { validationSchema } from "./validation";
import * as Yup from "yup";
import { useAuth } from "hooks/Auth";
import { useNavigation } from "@react-navigation/native";

interface IFormData {
  email: string;
  password: string;
}

const Login = function Login(): JSX.Element {
  const formRef = useRef<FormHandles>(null);
  const [modalVisible, setModalVisible] = useState(false);
  const { signIn } = useAuth();
  const navigation = useNavigation<any>();

  async function login(loginData: IFormData): Promise<string> {
    try {
      const response = await Api.post("/login", loginData);
      return response.data;
    } catch (erro) {
      console.log("erro", erro);
    }
    setModalVisible(true);
    return "";
  }

  async function handleSubmit() {
    try {
      formRef?.current?.setErrors({});

      const loginData: IFormData = {
        email: formRef.current?.getData().email,
        password: formRef.current?.getData().password,
      };

      await validationSchema.validate(loginData, {
        abortEarly: false,
      });

      const token: string = await login(loginData);
      signIn(token);
    } catch (errors) {
      const validationErrors = {} as any;
      if (errors instanceof Yup.ValidationError) {
        errors.inner.forEach((error) => {
          if (error.path) {
            validationErrors[error.path] = error.message;
          }
        });
        formRef?.current?.setErrors(validationErrors);
      }
    }
  }

  return (
    <>
      <Screen>
        <HeaderContainer>
          <ImageHeaderContainer>
            <Image
              source={require("../../assets/images/logo.png")}
              style={{ height: 120, width: 120 }}
            />
          </ImageHeaderContainer>
          <TitleScreen>Login</TitleScreen>
        </HeaderContainer>
        <BodyContainer>
          <Form ref={formRef} onSubmit={handleSubmit}>
            <TextInput
              leftIcon="user-o"
              name="email"
              label="Digite seu e-mail"
              placeholder="E-mail"
            />
            <PasswordInput
              leftIcon="lock"
              canRevealPassword={true}
              name="password"
              label="Digite sua senha"
              placeholder="Digite sua senha"
              autoCapitalize="none"
              textContentType="password"
            />
            <Button
              style={{ marginTop: 25 }}
              title="Enviar"
              onPress={() => {
                handleSubmit();
              }}
            />
          </Form>
          <TouchableOpacity>
            <ForgotPasswordText
              onPress={() => {
                navigation.navigate("ForgotPassword");
              }}
            >
              Esqueci minha senha!
            </ForgotPasswordText>
          </TouchableOpacity>
        </BodyContainer>
      </Screen>
      {modalVisible && (
        <Modal
          setModalVisible={setModalVisible}
          modalVisible={modalVisible}
        ></Modal>
      )}
    </>
  );
};

export default Login;
