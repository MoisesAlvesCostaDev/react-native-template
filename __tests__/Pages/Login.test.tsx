import "react-native";
import React from "react";
import Login from "pages/Login";
import { fireEvent, render } from "@testing-library/react-native";


jest.mock(
  'react-native-vector-icons/FontAwesome',
  () => 'MockedFontAwesome',
);

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
    }),
  };
});


describe("test render Login screen", () => {
  it("renders correctly", () => {
    const { getByTestId, getByText, toJSON } = render(
      <Login  />
    );
    getByText("Login");
    expect(toJSON()).toMatchSnapshot();
  });
});

// describe("test render and click button component", () => {
//   it("renders correctly", () => {
//     const onClick = jest.fn();
//     const { getByTestId } = render(<Button title="button" onPress={onClick} />);
//     const buttonStyled = getByTestId("buttonStyled");
//     fireEvent.press(buttonStyled);
//     expect(onClick).toHaveBeenCalledTimes(1);
//   });
// });
