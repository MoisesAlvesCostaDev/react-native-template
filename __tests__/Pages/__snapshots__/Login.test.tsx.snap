// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test render Login screen renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FF5C00",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 4,
          "flexShrink": 1,
          "paddingBottom": 13,
          "paddingLeft": 13,
          "paddingRight": 13,
          "paddingTop": 13,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/images/logo.png",
          }
        }
        style={
          Object {
            "height": 120,
            "width": 120,
          }
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontWeight": "bold",
          },
        ]
      }
    >
      Login
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderTopLeftRadius": 30,
          "borderTopRightRadius": 30,
          "flexBasis": 0,
          "flexGrow": 6,
          "flexShrink": 1,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 87,
        },
      ]
    }
  >
    <View>
      <View
        style={
          Array [
            Object {
              "height": 73,
              "marginBottom": 2,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#646464",
                "fontSize": 12,
              },
            ]
          }
        >
          Digite seu e-mail
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderBottomColor": "#646464",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "height": 40,
              },
            ]
          }
        >
          <MockedFontAwesome
            name="user-o"
            size={15}
            style={
              Object {
                "color": "#646464",
                "marginRight": 8,
              }
            }
          />
          <TextInput
            defaultValue=""
            onChangeText={[Function]}
            placeholder="E-mail"
            style={
              Object {
                "flex": 0.98,
                "paddingBottom": 0,
                "paddingTop": 0,
                "textAlignVertical": "center",
              }
            }
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#EE0000",
                "fontSize": 12,
                "textAlign": "right",
              },
            ]
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "height": 73,
              "marginBottom": 2,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#646464",
                "fontSize": 12,
              },
            ]
          }
        >
          Digite sua senha
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderBottomColor": "#646464",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "height": 40,
              },
            ]
          }
        >
          <MockedFontAwesome
            name="lock"
            size={20}
            style={
              Object {
                "color": "#646464",
                "marginRight": 8,
              }
            }
          />
          <TextInput
            autoCapitalize="none"
            defaultValue=""
            onChangeText={[Function]}
            placeholder="Digite sua senha"
            secureTextEntry={true}
            style={
              Object {
                "flex": 0.98,
                "paddingBottom": 0,
                "paddingTop": 0,
                "textAlignVertical": "center",
              }
            }
            textContentType="password"
          />
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <MockedFontAwesome
              name="eye"
              size={20}
              style={
                Object {
                  "color": "#646464",
                  "marginRight": 8,
                }
              }
            />
          </View>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#EE0000",
                "fontSize": 12,
                "textAlign": "right",
              },
            ]
          }
        />
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#EDEDED",
            "borderRadius": 12,
            "height": 48,
            "justifyContent": "center",
            "marginTop": 25,
            "opacity": 1,
          }
        }
        testID="buttonStyled"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#646464",
                "fontSize": 16,
              },
            ]
          }
        >
          Enviar
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        onPress={[Function]}
        style={
          Array [
            Object {
              "color": "#646464",
              "marginTop": 25,
              "textAlign": "center",
              "textDecorationLine": "underline",
            },
          ]
        }
      >
        Esqueci minha senha!
      </Text>
    </View>
  </View>
</View>
`;
